src/parallel_queue.rs

```diff
@@ -0,0 +1,290 @@
+            spare_cmds,
+            in_flight: VecDeque::new(),
+            active: None,
+        }
+    }
+}
+
+pub struct Handle {
+    send: mpsc::Sender<Work>,
+    shared: Arc<Shared>,
+    cmd_pool: vk::CommandPool,
+    spare_cmds: Vec<vk::CommandBuffer>,
+    in_flight: VecDeque<Work>,
+
+    active: Option<Work>,
+}
+
+impl Handle {
+    /// # Safety
+    /// `device` must match that passed to [`ParallelQueue::new`] and no work may be in flight
+    pub unsafe fn destroy(&mut self, device: &Device) {
+        device.destroy_command_pool(self.cmd_pool, None);
+    }
+
+    #[inline]
+    pub fn semaphore(&self) -> vk::Semaphore {
+        self.shared.semaphore
+    }
+
+    #[inline]
+    pub fn cmd_pool(&self) -> vk::CommandPool {
+        self.cmd_pool
+    }
+
+    /// Begin recording commands
+    ///
+    /// Must be called before calls to `cmd()` or `time()`. `end()` must be called soon after to
+    /// avoid stalling the queue.
+    ///
+    /// # Safety
+    /// `device` must match that passed to [`ParallelQueue::new`]
+    pub unsafe fn begin(&mut self, device: &Device) {
+        if self.active.is_some() {
+            return;
+        }
+        let time = NonZeroU64::new_unchecked(self.shared.next_time.fetch_add(1, Ordering::Relaxed));
+        let cmd = match self.spare_cmds.pop() {
+            Some(cmd) => cmd,
+            None => {
+                let complete = device
+                    .get_semaphore_counter_value(self.shared.semaphore)
+                    .unwrap();
+                while self
+                    .in_flight
+                    .front()
+                    .map_or(false, |work| work.time.get() <= complete)
+                {
+                    self.spare_cmds
+                        .push(self.in_flight.pop_front().unwrap().cmd);
+                }
+                if self.spare_cmds.is_empty() {
+                    self.spare_cmds.extend(
+                        device
+                            .allocate_command_buffers(
+                                &vk::CommandBufferAllocateInfo::builder()
+                                    .command_pool(self.cmd_pool)
+                                    .command_buffer_count(32),
+                            )
+                            .unwrap(),
+                    );
+                }
+                self.spare_cmds.pop().unwrap()
+            }
+        };
+        device
+            .begin_command_buffer(
+                cmd,
+                &vk::CommandBufferBeginInfo::builder()
+                    .flags(vk::CommandBufferUsageFlags::ONE_TIME_SUBMIT),
+            )
+            .unwrap();
+        self.active = Some(Work { cmd, time });
+    }
+
+    /// Whether `begin` has been called more recently than `end`
+    #[inline]
+    pub fn is_active(&self) -> bool {
+        self.active.is_some()
+    }
+
+    /// Command buffer to record commands on
+    #[inline]
+    pub fn cmd(&self) -> vk::CommandBuffer {
```

Comment:
I think it it would be nice to add a comment that this should only be called if `is_active()` is true and will panic otherwise.